Chap_API_Publish.tex

```diff
@@ -557,29 +557,32 @@ \section{Retrieval rules for published data}
 \endgroup
 
 The retrieval rules for published data primarily revolve around enforcing data access permissions and range constraints.
-All publish and lookup operations operate on a range as they default to \refconst{PMIX_RANGE_SESSION}.
-The key being looked up will match with a previously published key only if all of the following conditions are met:
+All publish and lookup operations operate on a range. If not specified, the range defaults to \refconst{PMIX_RANGE_SESSION}.
+The key being looked up will match with a published key only if all of the following conditions are met:
 
 \begin{enumerate}
-    \item The lookup key matches the previously published key
-    \item The publishing process of the previously published key falls within the lookup range specified by the requester.
+    \item The lookup key matches the published key.
+    \item The publisher of the published key falls within the lookup range specified by the requester.
     \item The the requester falls within the range specified by the publisher.
     \item If the publisher specified access permissions, the effective \ac{UID} and \ac{GID} of the requester must meet those requirements.
 \end{enumerate}
 
-Fig.~\ref{fig:publish_lookup} shows process P0 executing on host X in namespace 1 in session A publishing the key "tree" on three ranges.  Publishing to different ranges is allowed.
+Fig.~\ref{fig:publish_lookup} shows process P0 executing on Host X in Namespace 1 in Session A publishing the key "tree" on three ranges.  Publishing to different ranges is allowed.
```

Comment:
I agree.  I was just trying to capture what I thought the document says currently, but I agree it is very complex to understand and to implement and it still has non-determinant cases (results that may vary based on implementation).   I'm all for making something that is as simple as possible both to understand and implement, even if that means it is not very rich in features.  Like you said, this is just meant to be a simple rendezvous mechanism.   I need to think more about what options we have and what we could do without breaking backward compatibility.  This is one of those cases where I don't think anyone is invested in the way things "work" now, but not sure how much liberty we should take on changing behavior.  

On another note, I would love to see the Publish API take 2 info arrays... one for the things to publish and one for modifiers to the call.  As it is now if I call Publish with an array of infos like:  

PMIX_RANGE=PMIX_RANGE_NAMESPACE
tree=oak
PMIX_RANGE=PMIX_RANGE_GLOBAL
tree=elm
PMIX_TIMEOUT=10

then we have several problems:  If the implementation doesn't handle or recognize PMIX_TIMEOUT then it will actually publish PMIX_TIMEOUT as a key since it doesn't recognize it as a directive.  Also, will it publish tree=oak at the namespace range and tree=elm at the Global range or will it publish only one of the values of tree at some range (either namespace or global)?   Things get really tricky when the things to publish are combined with the directives.

But deprecating existing API's for new Publish API's seems like a lot of work for something that very few products are using and probably have figured out a way to not get tripped up by these things already.  I don't want the pmix-standard to be a burden to implementations or progress, but I would like it to describe how things actually work so someone can develop new clients without getting lost or confused.   I'll see if I can think up some way to simplify what is supported without breaking backward compatibility.

Reply:
So don't deprecate the API, but add an attribute for specifying modifiers to the call. It would take a `pmix_data_array_t` containing an array of `pmix_info_t` that has the modifiers in them. I don't think that would be much of a problem for backwards compatibility.

If you prefer, you could add an attribute for the things to publish, leaving all else in the call to be modifiers. Either way should be fine.

I wouldn't worry too much about changes in behavior at this point. OMPI is the primary user of it (for connect/accept), and I've already outlined a way for them to use `PMIx_Group` instead - which is a far better and more performant solution for that use-case.

Reply:
I like that idea.   We could have both approaches for a while, then move towards deprecating the old way after a while.

Reply:
Thinking some more, we should probably add the attribute for the things to publish (as opposed to the modifiers) so we remain consistent with the other APIs - whose info arrays all contain modifiers.